pyfr/integrators/dual/controllers.py

```diff
@@ -68,35 +69,50 @@ def advance_to(self, t):
                 # Activate convergence monitoring after pseudo-niters-min
                 if i >= self._minniters - 1:
                     # Subtract the current solution from the previous solution
-                    self._add(-1.0/dtau, self._idxprev, 1.0/dtau, self._idxcurr)
+                    self._add(-1.0, self._idxprev, 1.0, self._idxcurr)
 
                     # Compute the normalised residual and check for convergence
-                    if self._resid(self._idxprev, self._idxcurr) < 1.0:
+                    resid = self._resid(dtau, self._idxprev)
+                    self.stepinfo.append((dt, 'accept', resid))
+
+                    if max(resid) < self._residtol:
                         break
 
             # Update the dual-time stepping banks (n+1 => n, n => n-1)
             self.finalise_step(self._idxcurr)
 
             # We are not adaptive, so accept every step
-            self._accept_step(dt, self._idxcurr)
+            self._accept_step(dt, self._idxcurr, resid)
 
     @memoize
     def _get_errest_kerns(self):
-        return self._get_kernels('errest', nargs=3, norm='uniform')
+        return self._get_kernels('errest', nargs=3, norm=self._norm)
 
-    def _resid(self, x, y):
+    def _resid(self, dtau, x):
         comm, rank, root = get_comm_rank_root()
 
         # Get an errest kern to compute the square of the maximum residual
         errest = self._get_errest_kerns()
 
         # Prepare and run the kernel
-        self._prepare_reg_banks(x, y, y)
-        self._queue % errest(self._pseudo_aresid, self._pseudo_rresid)
-
-        # Reduce locally (element types) and globally (MPI ranks)
-        rl = max(errest.retval)
-        rg = comm.allreduce(rl, op=get_mpi('max'))
-
-        # Normalise
-        return math.sqrt(rg)
+        self._prepare_reg_banks(x, x, x)
+        self._queue % errest(dtau, 0.0)
+
+        # L2 norm
+        if self._norm == 'l2':
+            # Reduce locally (element types) and globally (MPI ranks)
+            rl = np.array([np.sum(e) for e in zip(*errest.retval)])
+            comm.Allreduce(get_mpi('in_place'), rl, op=get_mpi('sum'))
+
+            # Normalise
+            res = np.sqrt(rl / self._gndofs)
+        # L^∞ norm
+        else:
+            # Reduce locally (element types) and globally (MPI ranks)
+            rl = np.max(list(errest.retval), axis=0)
```

Comment:
This and line 104 is still ugly.  Start with `np.array(list(errest.retval))` and go from there.